### ISSUE TITLE
False return when tip occurs during I2C reading

### ISSUE THREAD

**Issue by aufdenkampe** [EnviroDIY/RainTipCounter#5]
In June @Volk3rJ noticed that his rain gauge returned a -9999 during a very intense rain event:

> Just a curious question. We seemingly had a pretty intense short rainfall event yestarday. Our LAF rain gauge read about 0.55” in 15 minutes (see first image below). Our McF gauge had 0.22” in one 5 minute intervall but a reported an outage (-9999) in the intervall before (second image below). Is it possible that the rainfall counts exceeded the max value in the trinket script or something like that? Just wondering since it just hapopened during a high intense event. Or was this likely just a coincedence?

He has let me know that this has happened a few other times.

**Comment by aufdenkampe** [comment]
@bschulz1701 responded:

> This is an interesting situation, what seems to have happened is that there was a tipping event being recorded at the same instant the data was requested. The way the modular sensors reading code is written, this could cause a problem. However, I think there should be a pretty easy solution.
> 
> Now for some more technical details:
> The problem occurs because when a tip occurs, the trinket is forced to deal with an operation for a short period of time and will not respond to the Mayfly. This is generally ok, because the trinket will simply respond very shortly after. However, this can become a problem based on how you ask for data from the trinket. The way this is done presently, the trinket is asked for data, and in that same line, they test to see if the information is received. Normally this is fine, but when there is an interrupt collision, there is a short delay between the command being received and the values being pushed out on the bus. Currently, as soon as the Mayfly finishes its request, it sends a stop command and stops talking on the bus. Which is a problem if the trinket has not responded yet, and results in the -9999 failure value you observed. 
> 
> I did some testing this morning, and it takes ~24us for the trinket to record a tip event and go back to waiting to send the data to the Mayfly. Using some of your numbers of ~0.55" in 15 minutes (which from my understanding is very heavy rain), and assuming we sample at some random interval within there, there is about a 1 in 625,000 chance of having a collision like that. Even going with normal heavy rain of 1"/hour, the likelihood is around 1 in 1.5 million samples, so very low likelihood, but of course, we don't want to just rely on chance. 
> 
> Potential solution:
> I believe it should be sufficient to change the call to the device in line 70 of the rain counter I2C library from 
```cpp
if (Wire.requestFrom(int(_i2cAddressHex),2))
```
> to
```cpp
if (Wire.requestFrom(int(_i2cAddressHex),2, false))    
```
> This will have the Mayfly send a restart command after the first transmit is completed. Generally this first transmit will succeed, but if it does not due to a collision, this should allow the Mayfly to still get the data a brief moment later. However, I have not tested this solution and it should be verified that not problems are caused during normal operation.
> 
> Another potential solution (which would require modifying the firmware on the trinket itself), would be to set it up such that the trinket does not clear the values from the register unless it successfully communicated with the Mayfly (in this case, you would still have a bad sample potentially, but your total number would be correct, instead of losing the data from that sample as well). I would need to think about exactly how to do this. Or, could change the firmware to have two functions instead of one, read the data, and clear the data. This way, the Mayfly could try to read, if it does, then send a clear message, otherwise just move along and wait for next time. This would result in the same kind of potential, where one point could be missing, but the total count would be correct.
> 
> The most complete hardware solution would be to switch over to a [Project Tally](https://github.com/NorthernWidget-Skunkworks/Project-Tally) event counter. This is a device I designed as a hardware solution to the trinket system, much more low power and more robust (since there are no potential colliding interrupts). Anthony has just gotten some of these for use with anemometers, but they would work equally well to replace the trinket event counter.

**Comment by aufdenkampe** [comment]
Yesterday, while I was working with @brianjastram, we were testing some new Trinket Pro's that we set as TippingBucketRainCounter slaves, and we could repeatedly get that collision that @bschulz1701 described, when using the [`MasterTest.ino`](https://github.com/EnviroDIY/TippingBucketRainCounter/tree/master/Master_Test) sketch from this repo, set with a 4 second interval.

Our results:
```
Number of Tips since last read = 1
Number of Tips since last read = 0
Number of Tips since last read = 65282
Number of Tips since last read = 65284
Number of Tips since last read = 65287
Number of Tips since last read = 3
Number of Tips since last read = 0
Number of Tips since last read = 0
```

It would be nice to find a firmware solution for the TippingBucketRainCounter devices.

**Comment by aufdenkampe** [comment]
I found the issue here!

Previously, the Slave was sending data by passing a 16-bit integer (2-bytes) to `Wire.write(value)` command like this:
https://github.com/EnviroDIY/TippingBucketRainCounter/blob/c66a4bae1571fe04fde6e23159120247fae1ff30/Slave_TippingBucket_Test_LowPower/Slave_TippingBucket_Test_LowPower.ino#L95

However, the `Wire.write(value)` function is designed to only receive 1-byte, so it was only passing 1 byte to the Master, and for some reason, sometimes when the first byte was somewhat large (>10 or so), the second byte would get read as `11111111` or 255, which is what is what is returned when there is no response and which would cause those unreasonable numbers.  

The multi-byte version, `Wire.write(data, length)`, requires the `data` input to be a byte-array with a specified, constant `length` in bytes. 

I implemented this in commit 11024f6 for a 4-byte integer (`uint32_t`), following the examples here: https://thewanderingengineer.com/2015/05/06/sending-16-bit-and-32-bit-numbers-with-arduino-i2c/#

This appears to have fixed the issue, as I can not reproduce the errors shown in https://github.com/EnviroDIY/TippingBucketRainCounter/issues/5#issuecomment-521409931 above, which had been previously easy to reproduce using the Master sketch.  

This fix also is a first step toward enabling the use of our TippingBucketRainGauge as a reed switch counter for an anemometer, as requested in #4.

This will fix will get rolled into the next release.

**Comment by aufdenkampe** [comment]
This bug has now been fully fixed, with updates to the I2C Mater code, both in the `Master_Test.ino` of this repo and also within the EnviroDIY Modular Sensors library with https://github.com/EnviroDIY/ModularSensors/commit/87cd0c995e5f6358fa3c11f25295d3a13e01c6fc.

Updating the firmware on the slave device is not required to fix this read error, as the update is backward compatible with either the old version of the slave (v0.1.0) or the update (>= v0.2.0).
